statistics/handle/lru_cache.go

```diff
@@ -0,0 +1,253 @@
+func (l *internalLRUCache) Del(key int64) {
+	var del bool
+	l.Lock()
+	del = l.del(key)
+	l.Unlock()
+	if del {
+		metrics.StatsCacheLRUCounter.WithLabelValues(typDel).Inc()
+	}
+}
+
+func (l *internalLRUCache) del(key int64) bool {
+	element := l.elements[key]
+	if element == nil {
+		return false
+	}
+	l.cache.Remove(element)
+	delete(l.elements, key)
+	l.cost -= element.Value.(*cacheItem).cost
+	return true
+}
+
+// Cost returns the current cost
+func (l *internalLRUCache) Cost() int64 {
+	l.RLock()
+	defer l.RUnlock()
+	return l.cost
+}
+
+// Keys returns the current Keys
+func (l *internalLRUCache) Keys() []int64 {
+	l.RLock()
+	defer l.RUnlock()
+	r := make([]int64, 0)
+	for _, v := range l.elements {
+		r = append(r, v.Value.(*cacheItem).key)
+	}
+	return r
+}
+
+// Values returns the current Values
+func (l *internalLRUCache) Values() []*statistics.Table {
+	l.RLock()
+	defer l.RUnlock()
+	r := make([]*statistics.Table, 0)
+	for _, v := range l.elements {
+		r = append(r, v.Value.(*cacheItem).value)
+	}
+	return r
+}
+
+// Len returns the current length
+func (l *internalLRUCache) Len() int {
+	l.RLock()
+	defer l.RUnlock()
+	return len(l.elements)
+}
+
+// CalculateCost re-calculate the memory message
+func (l *internalLRUCache) CalculateCost() {
+	var cost int64
+	l.Lock()
+	for _, v := range l.elements {
+		item := v.Value.(*cacheItem)
+		oldCost := item.cost
+		newCost := item.value.MemoryUsage()
+		item.cost = newCost
+		l.cost += newCost - oldCost
+	}
+	l.evictIfNeeded()
+	cost = l.cost
+	l.Unlock()
+	metrics.StatsCacheMemUsage.WithLabelValues("cache").Set(float64(cost))
+}
+
+// CalculateTableCost re-calculate the memory message for the certain key
+func (l *internalLRUCache) CalculateTableCost(key int64) {
+	var cost int64
+	var calculated bool
+	l.Lock()
+	calculated = l.calculateTableCost(key)
+	cost = l.cost
+	l.Unlock()
+	if calculated {
+		metrics.StatsCacheMemUsage.WithLabelValues("cache").Set(float64(cost))
+	}
+}
+
+func (l *internalLRUCache) calculateTableCost(key int64) bool {
+	element, exists := l.elements[key]
+	if !exists {
+		return false
+	}
+	item := element.Value.(*cacheItem)
+	l.put(item.key, item.value, item.value.MemoryUsage(), true)
+	l.evictIfNeeded()
+	return true
+}
+
+// Copy returns a replication of LRU
+func (l *internalLRUCache) Copy() kvCache {
+	var newCache *internalLRUCache
+	l.RLock()
+	newCache = newInternalLRUCache(l.capacity)
+	node := l.cache.Back()
+	for node != nil {
+		key := node.Value.(*cacheItem).key
+		value := node.Value.(*cacheItem).value
+		cost := node.Value.(*cacheItem).cost
+		newCache.put(key, value, cost, false)
+		node = node.Prev()
+	}
+	l.RUnlock()
+	metrics.StatsCacheLRUCounter.WithLabelValues(typCopy).Inc()
+	return newCache
+}
+
+func (l *internalLRUCache) evictIfNeeded() {
+	for l.cost > l.capacity {
+		lru := l.cache.Back()
+		droppedItem := lru.Value.(*cacheItem)
+		delete(l.elements, droppedItem.key)
+		l.cache.Remove(lru)
```

Comment:
Better to keep simple stats for table& column like rowCount, nullCount, ndv，but drop mem-consuming structure like buckets, cmsketch.